controller/server_controller.py

```diff
@@ -14,6 +16,62 @@
     parse_pod_metrics,
 )
 
+total_launch = Counter("sessions_launch_total", "Number of sessions launched")
+total_failed = Counter(
+    "sessions_failed_total", "Number of sessions which have failed being launched"
+)
+num_of_sessions = Gauge("sessions_all", "Number of sessions")
+num_of_running_sessions = Gauge(
+    "sessions_running", "Number of sessions in running state"
+)
+num_of_starting_sessions = Gauge(
+    "sessions_starting",
+    "Number of sessions which have been scheduled but are still starting",
+)
+
+
+class ServerStatusEnum(Enum):
+    """Simple Enum for server status."""
+
+    Running = "running"
+    Starting = "starting"
+    Stopping = "stopping"
+    Failed = "failed"
+
+    @classmethod
+    def list(cls):
+        return list(map(lambda c: c.value, cls))
+
+
+sessions_state = dict()
```

Comment:
@vigsterkr you do not need to save a state like this. All the state you need is in the k8s manifests already. And you get this in every handler in kopf automatically.

- for `sessions_launch_total` increment in the existing kopf.on.create below or make another one, never decrement
- for `sessions_all` increment in kopf.on.create, add a new kopf.on.delete handler and decrement there
- for `sessions_starting` increment in kopf.on.create, decrement whenever the state changes from running to something else
- for `sessions_running` increment when status changes from starting to running, decrement when status changes from running to anything else
- for `sessions_failed_total` increment when status goes to failed (from any previous status), never decrement

To compare the old and new status you can simply register a new field handler like this on the jupyterserver resources:
https://kopf.readthedocs.io/en/stable/handlers/#field-handlers. This will give you direct access to the new and old values under `status.state`.

Reply:
Also please note that `update_status` will run for every child of the jupyterServer. (see https://github.com/SwissDataScienceCenter/amalthea/pull/145/files#diff-6346752d3f2c1f6b94bc66266fc9a17eb6efc937693109761d26759d9d0ab5d6R425). But to update the state you need to look at only the `mainPod`. So you do not need to run the checks with every invocation because in this cases the `body` could be anything.

There is already a variable to tell you if you are handling the main pod or not [here](https://github.com/SwissDataScienceCenter/amalthea/pull/145/files#diff-6346752d3f2c1f6b94bc66266fc9a17eb6efc937693109761d26759d9d0ab5d6R425). So just run the state update when `is_main_pod` is true. In all other cases the handler triggered on something that (for now at least) we do not care about. These other children are services, ingresses, secretes, pvc, configmaps.